cmd/autoconnector.go

```diff
@@ -0,0 +1,310 @@
+// Copyright 2020-2021 JackTrip Labs, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/xthexder/go-jack"
+)
+
+const (
+	// CHANNELS is the default number of channels per client
+	CHANNELS = 2
+	// JT_* are pattern-matching tokens used to discover JACK ports
+	JT_RECEIVE    = ":receive_"
+	JT_SEND       = ":send_"
+	JT_RECEIVE_RX = ".*:receive_.*"
+	JT_SEND_RX    = ".*:send_.*"
+	// SC_* are pattern-matching tokens used to find SuperCollider ports
+	SC_IN  = "SuperCollider:in_"
+	SC_OUT = "SuperCollider:out_"
+	// SN_* are pattern-matching tokens used to find supernova ports
+	SN_IN  = "supernova:input_"
+	SN_OUT = "supernova:output_"
+	// JAMULUS_* are pattern-matching tokens used to find Jamulus ports
+	JAMULUS_INPUT_LEFT   = "Jamulus:input left"
+	JAMULUS_INPUT_RIGHT  = "Jamulus:input right"
+	JAMULUS_OUTPUT_LEFT  = "Jamulus:output left"
+	JAMULUS_OUTPUT_RIGHT = "Jamulus:output right"
+)
+
+// registrationChan is a singleton channel used for notifications
+var registrationChan chan jack.PortId = make(chan jack.PortId)
+
+type AutoConnector struct {
+	Name         string
+	Channels     int
+	JackClient   *jack.Client
+	ClientLock   sync.Mutex
+	KnownClients map[string]int
+}
+
+func NewAutoConnector() *AutoConnector {
+	return &AutoConnector{
+		Name:         "autoconnector",
+		Channels:     CHANNELS,
+		KnownClients: map[string]int{"Jamulus": 0},
+	}
+}
+
+// handlePortRegistration signals the notification channel when a new port is registered
+// NOTE: We cannot modify ports in the callback thread so use a channel
+func handlePortRegistration(port jack.PortId, register bool) {
+	if register {
+		registrationChan <- port
```

Comment:
this wasn't so easy to do in jack_autoconnect because it's a standalone service and intended to be more general purpose, for example supporting starting it up after connecting everything to jack. i made it more efficient than O(logn) in the most recent update by adding some lock-free checks to reduce unnecessary triggers. but i don't think that is necessary here. one of the advantages of moving the logic into here is that we can make more assumptions about the order of things.

Reply:
I hadn't thought of that, it looks like there are some methods exposed to turn port id -> port. I'll play around with that and see if we can use it.